File: course/string/P.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Locale;
import java.util.StringTokenizer;

public class P {

    private void solve() {
        int t = readInt();
        for (int test = 0; test < t; ++test) {
            solveTest();
        }
    }

    private void solveTest() {
        readInt(); // n
        int k = readInt();
        char[] s = readString().toCharArray();

        boolean can = getAnswer(s, k);
        yesNo(can);
    }

    private boolean getAnswer(char[] s, int k) {
        char[] values = new char[k];

        final char unknown = '?';
        Arrays.fill(values, unknown);

        for (int i = 0; i < s.length; ++i) {
            char actual = s[i];
            if (unknown == actual) continue;

            int group = i % k;
            char expected = values[group];

            if (unknown == expected) {
                values[group] = actual;
            } else if (expected != actual) {
                return false;
            }
        }

        final char zero = '0', one = '1';

        int balance = 0;
        for (char value : values) {
            if (zero == value) --balance;
            if (one == value) ++balance;
        }

        for (char value : values) {
            if (unknown == value) {
                if (balance < 0) {
                    ++balance;
                } else {
                    --balance;
                }
            }
        }

        return balance == 0;
    }

    //////////////////////////////////////////////////////////////////

    long sqrtLong(long x) {
        long root = (long)Math.sqrt(x);
        while (root * root > x) --root;
        while ((root + 1) * (root + 1) <= x) ++root;
        return root;
    }

    //////////////////////////////////////////////////////////////////

    private boolean yesNo(boolean yes) {
        return yesNo(yes, "YES", "NO");
    }

    private boolean yesNo(boolean yes, String yesString, String noString) {
        out.println(yes ? yesString : noString);
        return yes;
    }

    //////////////////////////////////////////////////////////////////

    private long readLong() {
        return Long.parseLong(readString());
    }

    private int[] readIntArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; ++i) a[i] = readInt();
        return a;
    }

    private int readInt() {
        return Integer.parseInt(readString());
    }

    private String readString() {
        while (!tok.hasMoreTokens()) {
            String nextLine = readLine();
            if (null == nextLine) return null;
            tok = new StringTokenizer(nextLine);
        }

        return tok.nextToken();
    }

    private String readLine() {
        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    //////////////////////////////////////////////////////////////////

    private BufferedReader in;
    private StringTokenizer tok;
    private PrintWriter out;

    private void initFileIO(String inputFileName, String outputFileName) throws FileNotFoundException {
        in = new BufferedReader(new FileReader(inputFileName));
        out = new PrintWriter(outputFileName);
    }

    private void initConsoleIO() throws IOException {
        in = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(System.out);
    }

    private void initIO() throws IOException {
        Locale.setDefault(Locale.US);

        String fileName = "";

        if (!fileName.isEmpty()) {
            initFileIO(fileName + ".in", fileName + ".out");
        } else {
            if (new File("input.txt").exists()) {
                initFileIO("input.txt", "output.txt");
            } else {
                initConsoleIO();
            }
        }

        tok = new StringTokenizer("");
    }

    //////////////////////////////////////////////////////////////////

    private void run() {
        try {
            long timeStart = System.currentTimeMillis();

            initIO();
            solve();
            out.close();

            long timeEnd = System.currentTimeMillis();
            System.err.println("Time(ms) = " + (timeEnd - timeStart));
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        new P().run();
    }
}

File: course/string/P.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using ld = long double;
using ii = pair<int, int>;

const char SPACE = ' ', ENDL = '\n';

//////////////////////////////////////////////////////////////////

template <typename T>
T read() {
    T value;
    cin >> value;
    return value;
}

int ri() {
    return read<int>();
}

ll rll() {
    return read<ll>();
}

string rs() {
    return read<string>();
}

template<typename T>
vector<T> rv(int size) {
    vector<T> a(size);
    for (int i = 0; i < size; ++i) {
        a[i] = read<T>();
    }
    return a;
}

//////////////////////////////////////////////////////////////////

bool yn(bool result, string yes = "YES", string no = "NO") {
	cout << (result ? yes : no) << ENDL;
	return result;
}

//////////////////////////////////////////////////////////////////

vector<string> split(string& s, string seps) {
    vector<string> parts;

    string part = "";
    auto add_part = [&]() {
        parts.push_back(part);
        part = "";
    };

    for (char ch : s) {
        if (string::npos != seps.find(ch)) {
            add_part();
        } else {
            part += ch;
        }
    }

    add_part();

    return parts;
}

//////////////////////////////////////////////////////////////////

ld sqrt_ld(ld value) {
    return sqrt(value);
}

//////////////////////////////////////////////////////////////////

void init_io() {
    string file_name = "";
    if (!file_name.empty()) {
        string input_file_name = file_name + ".in";
        freopen(input_file_name.c_str(), "rt", stdin);

        string output_file_name = file_name + ".out";
        freopen(output_file_name.c_str(), "wt", stdout);
    } else {
        #ifndef ONLINE_JUDGE
        string input_file_name = "../../input.txt"; // reading from file in parent of parent of directory with file
        freopen(input_file_name.c_str(), "rt", stdin);
        #endif
    }

	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	cout << setprecision(10) << fixed;
}

bool get_ans(string& s, int k) {
    const char unknown = '?';
    vector<char> values(k, unknown);

    for (int i = 0; i < s.length(); ++i) {
        char actual = s[i];
        if (unknown == actual) continue;

        int group = i % k;
        char expected = values[group];

        if (unknown == expected) {
            values[group] = actual;
        } else if (expected != actual) {
            return false;
        }
    }

    const char zero = '0', one = '1';

    int balance = 0;
    for (char value : values) {
        if (zero == value) --balance;
        if (one == value) ++balance;
    }

    for (char value : values) {
        if (unknown == value) {
            if (balance < 0) ++balance;
            else --balance;
        }
    }

    return balance == 0;
}

void solve_test() {
    ri(); // n
    int k = ri();
    string s = rs();

    bool can = get_ans(s, k);
    yn(can);
}

void solve() {
    int t = ri();
    for (int test = 0; test < t; ++test) {
        solve_test();
    }
}

int main() {
    init_io();
    solve();
    return 0;
}